src/fetchers/git/default.nix

```diff
@@ -18,14 +23,21 @@ in {
     rev,
     ...
   } @ inp:
-    if b.match "refs/(heads|tags)/.*" rev == null && builtins.match "[a-f0-9]*" rev == null
+    if isGitRef rev == null && isGitRev rev == null
     then throw ''rev must either be a sha1 revision or "refs/heads/branch-name" or "refs/tags/tag-name"''
+    else if isGitRef (inp.ref or "") == null
+    then throw ''ref must be in either "refs/heads/branch-name" or "refs/tags/tag-name" format''
     else let
       b = builtins;
 
       refAndRev =
-        if b.match "refs/(heads|tags)/.*" inp.rev != null
+        # if the source specifies a ref, then we add both the ref and rev
+        if inp ? ref
+        then {inherit (inp) rev ref;}
+        # otherwise check if the rev is a ref, if it is add to ref
+        else if b.match "refs/(heads|tags)/.*" inp.rev != null
```

Comment:
`isGitRef inp.rev` is computed a second time here. Maybe use let..in to store the result?